Arcbox (#503)
microsoft/azure_arc: docs/azure_jumpstart_arcbox/_index.md

# Jumpstart ArcBox - An Azure Arc Jumpstart project

## Jumpstart ArcBox - Overview

![ArcBox architecture diagram](./arch.png)

Jumpstart ArcBox is a project that provides an easy to deploy sandbox for all things Azure Arc. ArcBox is designed to be completely self-contained within a single Azure subscription and resource group, which will make it easy for a user to get hands-on with all available Azure Arc technology with nothing more than an available Azure subscription.

> 

Reviewer: Make consider adding short use-case section?

## Azure Arc capabilities available in ArcBox

### Azure Arc enabled Servers

![ArcBox servers diagram](./servers.png)

> ArcBox includes three Azure Arc enabled server resources. These servers are hosted using nested virtualization in Azure. As part of the deployment, a Hyper-V host (ArcBox-Client) is deployed with three guest virtual machines. These machines, _ArcBoxWin_, _ArcBoxUbuntu_, and _ArcBoxSQL_ are connected as Azure Arc enabled servers via the ArcBox automation.

Reviewer: ```suggestion
ArcBox includes three Azure Arc enabled server resources and are hosted using nested virtualization in Azure. As part of the deployment, a Hyper-V host (ArcBox-Client) is deployed with three guest virtual machines. These machines, _ArcBoxWin_, _ArcBoxUbuntu_, and _ArcBoxSQL_ are connected as Azure Arc enabled servers via the ArcBox automation.
```

### Azure Arc enabled Kubernetes

![ArcBox Kubernetes diagram](./k8s.png)

ArcBox deploys one single-node Rancher K3s cluster running on an Azure virtual machine. This cluster is then connected to Azure as an Azure Arc enabled Kubernetes resource (_ArcBox-K3s_).

### Azure Arc enabled Data Services

> ArcBox deploys an AKS cluster, which is then used to deploy an Azure Arc enabled data services data controller.

Reviewer: ```suggestion
ArcBox deploys an AKS cluster, which is then being used to deploy an Azure Arc enabled data services data controller.
```

![ArcBox data services diagram](./dataservices.png)

## Automation flow

![Deployment flow diagram](./deploymentflow.png)

> ArcBox uses an advanced automation flow to deploy and configure all necessary resources with minimal user interaction. The above diagram provides a high level overview of the deployment flow. A high level summary of the deployment is:

Reviewer: ```suggestion
ArcBox uses an advanced automation flow to deploy and configure all necessary resources with minimal user interaction. The above diagram provides a high-level overview of the deployment flow. A high-level summary of the deployment is:
```

* User deploys the primary ARM template (azuredeploy.json). This template contains several nested templates that will run simultaneously.

>   * ClientVM ARM template - deploys the Client Windows VM

Reviewer: ```suggestion
  * ClientVM ARM template - deploys the Client Windows VM. This is the Hyper-V host VM where all user interactions with the environment are made from. 
```

  * AKS ARM template - deploys AKS cluster which will be used to run Azure Arc enabled data services

>   * Rancher K3s template - deploys a Linux VM which will have Rancher (K3s) installed on it and connected as an Azure Arc enabled Kubernetes cluster

Reviewer: ```suggestion
  * Rancher K3s template - deploys a Linux, Ubuntu VM which will have Rancher (K3s) installed on it and connected as an Azure Arc enabled Kubernetes cluster
```

  * Storage account template - used for staging files in automation scripts

>   * Management artifacts template - deploys Log Analytics workspace and solutions and Azure Policy artifacts

Reviewer: ```suggestion
  * Management artifacts template - deploys Azure Log Analytics workspace and solutions and Azure Policy artifacts
```

* User remotes into Client Windows VM, which automatically kicks off multiple scripts that:
  * Deploy and configure three (3) nested virtual machines in Hyper-V
    * Windows VM - onboarded as Azure Arc enabled Server
    * Ubuntu VM - onboarded as Azure Arc enabled Server

>     * Windows VM running SQL Server - onboarded as Azure Arc enabled SQL Server

Reviewer: ```suggestion
    * Windows VM running SQL Server - onboarded as Azure Arc enabled SQL Server (as well as Azure Arc enabled Server)
```

>   * Deploy and configure Azure Arc enabled data services on the AKS cluster including a data controller, a SQL MI instance, and a Postgres instance. After deployment, Azure Data Studio opens automatically with connection entries for each database instance. Data services deployed by the script are:

Reviewer: ```suggestion
  * Deploy and configure Azure Arc enabled data services on the AKS cluster including a data controller, a SQL MI instance, and a PostgreSQL Hyperscale cluster. After deployment, Azure Data Studio opens automatically with connection entries for each database instance. Data services deployed by the script are:
```

    * Data controller
    * SQL MI instance

>     * Postgres instance

Reviewer: This part feels repetitive

## Deployment Option 1: Azure Portal

* Click the [![Deploy to Azure](https://aka.ms/deploytoazurebutton)](https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2Fmicrosoft%2Fazure_arc%2Farcbox%2Fazure_jumpstart_arcbox%2Fazuredeploy.json) button and enter values for the the ARM template parameters.

> 

Reviewer: Add a screenshot of the Azure portal with the parameters view. I also recommend adding the "Deployment in progress" screen and "deployment completed" screen.

  ```shell
  az --version
  ```

* [Generate SSH Key](https://help.github.com/articles/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent/) (or use existing ssh key)

* Create Azure service principal (SP)

    To deploy ArcBox an Azure service principal assigned with the "Contributor" role is required. To create it login to your Azure account run the below command (this can also be done in [Azure Cloud Shell](https://shell.azure.com/)).

    ```shell
    az login
    az ad sp create-for-rbac -n "<Unique SP Name>" --role contributor
    ```

    For example:

    ```shell
    az ad sp create-for-rbac -n "http://AzureArcBox" --role contributor
    ```

    Output should look like this:

    ```json
    {
    "appId": "XXXXXXXXXXXXXXXXXXXXXXXXXXXX",
    "displayName": "AzureArcBox",
    "name": "http://AzureArcBox",
    "password": "XXXXXXXXXXXXXXXXXXXXXXXXXXXX",
    "tenant": "XXXXXXXXXXXXXXXXXXXXXXXXXXXX"
    }
    ```

    > **Note: It is optional but highly recommended to scope the SP to a specific [Azure subscription and resource group](https://docs.microsoft.com/en-us/cli/azure/ad/sp?view=azure-cli-latest)**

* Login to AZ CLI using the ```az login``` command.

> * Now you will need to edit the [azuredeploy.parameters.json](../../azure_jumpstart_arcbox/azuredeploy.parameters.json) ARM template parameters file and supply some values for your environment.

Reviewer: ```suggestion
* Edit the [azuredeploy.parameters.json](../../azure_jumpstart_arcbox/azuredeploy.parameters.json) ARM template parameters file and supply some values for your environment.
```

  * *aksDnsPrefix* - AKS unique DNS prefix
  * *sshRSAPublicKey* - Your SSH public key
  * *spnClientId* - Your Azure service principal id
  * *spnClientSecret* - Your Azure service principal secret
  * *spnTenantId* - Your Azure tenant id
  * *windowsAdminUsername* - Client Windows VM Administrator name
  * *windowsAdminPassword* - Client Windows VM administrator password
  * *myIpAddress* - Your local IP address. This is used to allow remote RDP and SSH connections to the Client Windows VM and K3s Rancher VM.
  * *logAnalyticsWorkspaceName* - Unique name for the ArcBox log analytics workspace

>     ![Screenshot showing example parameters](./parameters.png)

Reviewer: make sure to nuke the SPN in the screenshot

Author: this is not a real key

* Now you will deploy the ARM template. Navigate to the local cloned [deployment folder](../../azure_jumpstart_arcbox) and run the below command:

  ```shell
  az group create --name <Name of the Azure resource group> --location <Azure Region>
  az deployment group create \
  --resource-group <Name of the Azure resource group> \
  --template-file azuredeploy.json \
  --parameters azuredeploy.parameters.json 
  ```

  ![Screenshot showing az group create](./azgroupcreate.png)

>   ![Screenshot showing az deployment group create](./azdeploy.png)

Reviewer: mask subscription

* After deployment, you should see the ArcBox resources inside your resource group.

  ![Screenshot showing az deployment group create](./deployedresources.png)

> * Open a remote desktop connection into ArcBox-Client. Upon logging in, multiple automated scripts will open and start running. These scripts will take 10-20 minutes to finish. Once completed, Azure Data Studio will start. At this point, the deployment is complete.

Reviewer: ```suggestion
* Open a remote desktop connection into _ArcBox-Client_. Upon logging in, multiple automated scripts will open and start running. These scripts will take 10-20 minutes to finish and once completed, Azure Data Studio will start. At this point, the deployment is complete.
```

  ![Screenshot showing ArcBox-Client](./clientscript.png)

> 

Reviewer: I think there is a "Usage" section missing to show how to use the environment:

- Details on how to login the nested VM
- SQL Management studio in the SQL VM
- SQL Assessment in the portal
- No mention on kubectx and how to interact with the AKS/K3s clusters
- Azure Monitor for containers on the K3s Arc cluster (And AKS for the matter)
- azdata login to the controller
- etc

> 

Reviewer: Add a screenshot and highlight the Arc resources

## Example ArcBox usage scenarios

> ArcBox is a sandbox that can be used for a large variety of use cases, such as a sandbox environment for testing and training or kickstarter for proof of concept projects. Some suggested next steps for you to try in your ArcBox are:

Reviewer: Consider referencing (hyperlinks) to specific scenarios section in the Jumpstart

> * Deploy GitOps configurations on the K8s clusters

Reviewer: ```suggestion
* Deploy GitOps configurations with Azure Arc enabled Kubernetes
```

* Build policy initiatives that apply to your Azure Arc enabled resources
* Incorporate your own tooling and automation into the existing automation framework
* Build a certificate/secret/key management strategy with your Azure Arc resources

> 

Reviewer: - Azure Policy?
- Key Vault?
- something else?

## Clean up the deployment

To clean up your deployment, simply delete the resource group.

```shell
az group delete -n <name of your resource group>
```

![Screenshot showing az group delete](./azdelete.png)

> 

Reviewer: add screenshot for Azure portal resource group delete

## Known issues

> * Sample databases are not deployed into the MSSQL or Postgres instances at this time.

Reviewer: ```suggestion
* Sample databases are not deployed into the MSSQL or PostgreSQL instances at this time.
```

# Jumpstart ArcBox - An Azure Arc Jumpstart project

## Jumpstart ArcBox - Overview

![ArcBox architecture diagram](./arch.png)

Jumpstart ArcBox is a project that provides an easy to deploy sandbox for all things Azure Arc. ArcBox is designed to be completely self-contained within a single Azure subscription and resource group, which will make it easy for a user to get hands-on with all available Azure Arc technology with nothing more than an available Azure subscription.

> Azure Arc capabilities available in ArcBox include:

Reviewer: ```suggestion
## Azure Arc capabilities available in ArcBox
```

### Azure Arc enabled Servers

![ArcBox servers diagram](./servers.png)

> ArcBox includes three Azure Arc enabled server resources. These servers are hosted using nested virtualization in Azure. As part of the deployment, a Hyper-V host (ArcBox-Client) is deployed with three guest virtual machines. These machines, ArcBox-Win, ArcBox-Ubuntu, and ArcBox-SQL are connected as Azure Arc enabled servers via the ArcBox automation.

Reviewer: ```suggestion
ArcBox includes three Azure Arc enabled server resources. These servers are hosted using nested virtualization in Azure. As part of the deployment, a Hyper-V host (ArcBox-Client) is deployed with three guest virtual machines. These machines, _ArcBoxWin_, _ArcBoxUbuntu_, and _ArcBoxSQL_ are connected as Azure Arc enabled servers via the ArcBox automation.
```

Suggest to keep current names until we will update

### Azure Arc enabled Kubernetes

![ArcBox Kubernetes diagram](./k8s.png)

> ArcBox deploys one Rancher (K3s) cluster running on an Azure virtual machine. This cluster is then connected to Azure as an Azure Arc enabled Kubernetes resource (ArcBox-K3s).

Reviewer: ```suggestion
ArcBox deploys one, single-node Rancher K3s cluster running on an Azure virtual machine. This cluster is then connected to Azure as an Azure Arc enabled Kubernetes resource (_ArcBox-K3s_).
```

### Azure Arc enabled Data Services

ArcBox deploys an AKS cluster, which is then used to deploy an Azure Arc enabled data services data controller.

![ArcBox data services diagram](./dataservices.png)

## Prerequisites

* Clone the Azure Arc Jumpstart repository

    ```shell
    git clone https://github.com/microsoft/azure_arc.git
    ```

* [Install or update Azure CLI to version 2.15.0 and above](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest). Use the below command to check your current installed version.

  ```shell
  az --version
  ```

* [Generate SSH Key](https://help.github.com/articles/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent/) (or use existing ssh key)

* Create Azure service principal (SP)

    To connect the GCP virtual machine to Azure Arc, an Azure service principal assigned with the "Contributor" role is required. To create it, login to your Azure account run the below command (this can also be done in [Azure Cloud Shell](https://shell.azure.com/)).

    ```shell
    az login
    az ad sp create-for-rbac -n "<Unique SP Name>" --role contributor
    ```

    For example:

    ```shell

>     az ad sp create-for-rbac -n "http://AzureArcGCP" --role contributor

Reviewer: ```suggestion
    az ad sp create-for-rbac -n "http://AzureArcBox" --role contributor
```

  * AKS ARM template - deploys AKS cluster which will be used to run Azure Arc enabled data services
  * Rancher K3s template - deploys an Ubuntu Linux VM which will have Rancher (K3s) installed on it and connected as an Azure Arc enabled Kubernetes cluster
  * Storage account template - used for staging files in automation scripts
  * Management artifacts template - deploys Azure Log Analytics workspace and solutions and Azure Policy artifacts
* User remotes into Client Windows VM, which automatically kicks off multiple scripts that:
  * Deploy and configure three (3) nested virtual machines in Hyper-V
    * Windows VM - onboarded as Azure Arc enabled Server
    * Ubuntu VM - onboarded as Azure Arc enabled Server
    * Windows VM running SQL Server - onboarded as Azure Arc enabled SQL Server (as well as Azure Arc enabled Server)
  * Deploy and configure Azure Arc enabled data services on the AKS cluster including a data controller, a SQL MI instance, and a PostgreSQL Hyperscale cluster. After deployment, Azure Data Studio opens automatically with connection entries for each database instance. Data services deployed by the script are:
    * Data controller
    * SQL MI instance

>     * Postgres instance

Reviewer: ```suggestion
    * PostgreSQL Hyperscale cluster
```

    ![Screenshot showing example parameters](./parameters.png)

* Now you will deploy the ARM template. Navigate to the local cloned [deployment folder](../../azure_jumpstart_arcbox) and run the below command:

  ```shell
  az group create --name <Name of the Azure resource group> --location <Azure Region>
  az deployment group create \
  --resource-group <Name of the Azure resource group> \
  --template-file azuredeploy.json \
  --parameters azuredeploy.parameters.json 
  ```

  ![Screenshot showing az group create](./azgroupcreate.png)

  ![Screenshot showing az deployment group create](./azdeploy.png)

* After deployment, you should see the ArcBox resources inside your resource group.

  ![Screenshot showing az deployment group create](./deployedresources.png)

* Open a remote desktop connection into _ArcBox-Client_. Upon logging in, multiple automated scripts will open and start running. These scripts will take 10-20 minutes to finish and once completed, Azure Data Studio will start. At this point, the deployment is complete.

  ![Screenshot showing ArcBox-Client](./clientscript.png)

  ![Screenshot showing ArcBox resources in Azure Portal](./rgarc.png)

> ## Using ArcBox

Reviewer: We need more explanations/details here on the tools installed on the client VM:

1. How does a user log in to the nested VMs? we need to provide the username/password in the guide.
2. How does a user switch between the AKS/k3s clusters using kubectx.
3. Show how to do _azdata login_ and the fact the CLI is there.

Without knowing these things they will have to dig in the scripts to figure it out.
